src/EditorFeatures/Core/Implementation/GoToImplementation/GoToImplementationCommandHandler.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Commands;
+using Microsoft.CodeAnalysis.Editor.Host;
+using Microsoft.CodeAnalysis.Editor.Navigation;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Notification;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.GoToImplementation
+{
+    [ExportCommandHandler(PredefinedCommandHandlerNames.GoToDefinition,
+        ContentTypeNames.RoslynContentType)]
+    internal sealed class GoToImplementationCommandHandler : ICommandHandler<GoToImplementationCommandArgs>
+    {
+        private readonly IWaitIndicator _waitIndicator;
+        private readonly IEnumerable<Lazy<INavigableItemsPresenter>> _navigatableItemPresenters;
+
+        [ImportingConstructor]
+        public GoToImplementationCommandHandler(
+            [ImportMany] IEnumerable<Lazy<INavigableItemsPresenter>> navigatableItemPresenters,
+            IWaitIndicator waitIndicator)
+        {
+            _navigatableItemPresenters = navigatableItemPresenters;
+            _waitIndicator = waitIndicator;
+        }
+
+        public CommandState GetCommandState(GoToImplementationCommandArgs args, Func<CommandState> nextHandler)
+        {
+            // Because this is expensive to compute, we just always say yes
+            return CommandState.Available;
+        }
+
+        public void ExecuteCommand(GoToImplementationCommandArgs args, Action nextHandler)
+        {
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+
+            if (caret.HasValue)
+            {
+                var document = args.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document != null)
+                {
+                    // We have all the cheap stuff, so let's do expensive stuff now
+                    string messageToShow = null;
+                    _waitIndicator.Wait(
+                        EditorFeaturesResources.GoToImplementationTitle,
```

Comment:
Cool, I was unaware that bug existed. David's bug is for INotificationService that I use a few lines below, not the wait indicator. We have plenty of features passing titles today, so I'd like to either:

1. Completely remove the title parameter, and remove all the strings so we're not spending the loc cost.
2. Update the VS implementation to not use the title in the message box description, but leave titles for other potential hosts.

This is really a bad behavior of IVsUIShell.ShowMessageBox, I think, which just decides to glob the text together for no good reason.